dusk/grammar.py

```diff
@@ -186,13 +199,17 @@ def location_type(self, name: str):
         return sir.LocationType.Value(name)
 
     @transform(Capture(list).to("py_stmts"))
-    def statements(self, py_stmts: t.List):
+    def statements(self, py_stmts: t.List, in_stencil_root_scope: bool = False):
```

Comment:
This is also a hack, just like `Scope.is_stencil_scope` was a hack. But I think this is still better because:

* It doesn't _spoil_ the interface of `Scope` (I think a scope shouldn't know about _stencil scopes_ due to the _separation of concerns_ principle).
* It will also disallow temporary field declarations inside of top level if/else statements (the previous implementation allowed that).
* It makes it trivial to know when it's a temporary field right now.
* The hack feels more _contained_ like this.

Nevertheless, when we want to add support for locals, we'll have to rethink a lot of that again. But I think that'll require a bigger restructuring anyways, since we'll have to look at the declaration type to decide whether it's a temporary field or local variable. In the future, we'll probably have to consider adding proper infrastructure for this feature to `semantics.py` anyways.

Reply:
* scope has a pointer to it's parent. While `Scope.is_stencil_scope` is a hack for sure, I feel that having a query that returns the "length" of this pointer chain would be very reasonable. However, I felt that would warrant a different PR and hence didn't introduce it.
* good catch on the top level if/else stmt!
* If you prefer this hack over the other one I'm fine with that

Reply:
Fair points. Let's see how this develops in the future.